File: src/types/FunctionModelVersion.ts
// Interfaces for FunctionModelVersion update and retrieval processes

/**
 * Interface for the request to read a FunctionModel's FunctionModelVersions.
 */
export interface ReadFunctionModelVersionsRequest {
  function_model_uuid: string;
}

/**
 * Interface for the request to read a FunctionModelVersion's information.
 */
export interface ReadFunctionModelVersionRequest {
  uuid: string;
}

/**
 * Interface for the request to update the published FunctionModelVersion.
 */
export interface UpdatePublishedFunctionModelVersionRequest {
  uuid: string; // FunctionModelVersion UUID
  project_uuid: string;
  project_version: number;
  previous_published_version_uuid: string;
}

/**
 * Interface for the request to update a FunctionModelVersion's tags.
 */
export interface UpdateFunctionModelVersionTagsRequest {
  uuid: string;
  tags: Array<string>;
}

/**
 * Interface for the request to update a FunctionModelVersion's memo.
 */
export interface UpdateFunctionModelVersionMemoRequest {
  uuid: string;
  memo: string;
}

/**
 * General interface for representing a FunctionModelVersion in the system.
 */
export interface FunctionModelVersion {
  id: number;
  created_at: string;
  uuid: string;
  version: number;
  model: string;
  is_published: true;
  from_version: number;
  parsing_type: string;
  output_keys: Array<string>;
  functions: Array<string>;
  is_ab_test: true;
  ratio: number;
  tags: Array<string>;
  memo: string;
  function_model_uuid: string;
}

File: backend/api/web/models/function_model_version.py
"""APIs for FunctionModelVersion"""
from typing import List, Optional

from modules.types import PMObject
from api.common.models import PromptConfig


class FunctionModelVersionInstance(PMObject):
    id: int
    created_at: str
    uuid: str

    version: int
    model: str
    is_published: bool

    from_version: Optional[int] = None
    parsing_type: Optional[str] = None
    output_keys: Optional[List[str]] = None

    functions: Optional[List[str]] = None
    is_ab_test: Optional[bool] = None
    ratio: Optional[float] = None
    tags: Optional[List[str]] = None
    memo: Optional[str] = None

    function_model_uuid: str
    created_by: Optional[str] = None


class FunctionModelVersionAuthor(PMObject):
    email: Optional[str] = None
    image_url: Optional[str] = None


class FunctionModelVersionWithUserInstance(FunctionModelVersionInstance):
    user: FunctionModelVersionAuthor


class CreateFunctionModelVersionBody(PMObject):
    project_uuid: str
    function_model_uuid: str
    prompts: List[PromptConfig]
    model: Optional[str] = "gpt-3.5-turbo"
    from_version: Optional[int] = None
    parsing_type: Optional[str] = None
    output_keys: Optional[List[str]] = None
    functions: Optional[List[str]] = None


class UpdatePublishedFunctionModelVersionBody(PMObject):
    project_uuid: str
    project_version: int
    previous_published_version_uuid: Optional[str] = None


class UpdateFunctionModelVersionTagsBody(PMObject):
    tags: Optional[List[str]] = None


class FunctionModelVersionBatchRunInstance(PMObject):
    uuid: str
    created_at: str
    finished_at: str
    dataset_uuid: str
    status: str
    score: Optional[float] = None
